dali/kernels/imgproc/convolution/cutlass/kernel/gemm.h

```diff
@@ -152,153 +184,209 @@ struct Gemm {
 
-    typename Mma::FragmentC accumulators;
+      // Problem size is a function of threadblock index in the K dimension
+      int problem_size_k =
+          min(params.problem_size.k(), (threadblock_tile_offset.k() + 1) * params.gemm_k_size);
+      int total_gemm_k_iterations = (problem_size_k + Mma::Shape::kK - 1) / Mma::Shape::kK;
 
-    accumulators.clear();
+      // where the data ends
+      int k_end_offset = min(problem_size_k - 1, conv_diag_position + tile_extent + right_span);
+      int k_last_iter = ((k_end_offset + Mma::Shape::kK - 1) / Mma::Shape::kK);
 
-    if (!kSplitKSerial || gemm_k_iterations > 0) {
-      // Compute threadblock-scoped matrix multiply-add
-      mma(gemm_k_iterations, accumulators, iterator_A, iterator_B, accumulators);
-    }
+      // this is how many iterations we need if we start at the offset
+      int start_iteration =
+          (problem_size_k - k_skipped_offset + Mma::Shape::kK - 1) / Mma::Shape::kK;
+      int end_iteration = max(total_gemm_k_iterations - k_last_iter - 1, 0);
 
-    //
-    // Epilogue
-    //
+      // Compute position within threadblock
+      int thread_idx = threadIdx.x;
 
-    OutputOp output_op(params.output_op);
+      auto window_smem = shared_storage.main_loop.operand_Window_ref(params.window_size);
 
-    //
-    // Masked tile iterators constructed from members
-    //
+      // Transfer window from GMEM to SMEM
+      transfer_conv_window(params, window_smem);
 
-    threadblock_tile_offset = threadblock_swizzle.get_tile_offset();
+      void *in_data = params.ref_In.data();
+      void *window_data = window_smem.data();
 
-    // assume identity swizzle
-    MatrixCoord threadblock_offset(threadblock_tile_offset.m() * Mma::Shape::kM,
-                                   threadblock_tile_offset.n() * Mma::Shape::kN);
+      // Construct iterators to A and B operands
+      // One is proper GMEM iterator, one generates the matrix on the fly from conv window
+      typename Mma::IteratorA iterator_A(
+          select_A<kIsInnerConv>(params.params_In, params.params_Window),
+          select_A<kIsInnerConv>(params.ref_In.data(), window_smem.data()),
+          {params.problem_size.m(), problem_size_k}, thread_idx, tb_offset_A);
 
-    int block_idx =
-        threadblock_tile_offset.m() + threadblock_tile_offset.n() * params.grid_tiled_shape.m();
+      typename Mma::IteratorB iterator_B(
+          select_B<kIsInnerConv>(params.params_In, params.params_Window),
+          select_B<kIsInnerConv>(params.ref_In.data(), window_smem.data()),
+          {problem_size_k, params.problem_size.n()}, thread_idx, tb_offset_B);
 
-    // Construct the semaphore.
-    Semaphore semaphore(params.semaphore + block_idx, thread_idx);
+      // Broadcast the warp_id computed by lane 0 to ensure dependent code
+      // is compiled as warp-uniform.
+      int warp_idx = __shfl_sync(0x1f, threadIdx.x / 32, 0);
```

Comment:
```suggestion
      int warp_idx = __shfl_sync(0x1f, threadIdx.x / 32, 0);
```
Why is the mask 31 and nof 0xffffffff? Are you sure just 5 threads take part in this operation?

Reply:
done